db/changes.go

```diff
@@ -26,32 +26,62 @@ import (
+	ActiveOnly     bool               // If true, only return information on non-deleted, non-removed revisions
+	Revocations    bool               // Specifies whether revocation messages should be sent on the changes feed
+	clientType     clientType         // Can be used to determine if the replication is being started from a CBL 2.x or SGR2 client
+	ChangesCtx     context.Context    // Used for cancelling checking the changes feed should stop
+	VersionType    ChangesVersionType // The type of version to use for the changes feed. This is used to determine whether to use send revtree IDs or CV in the changes feed entries.
+}
+
+// ChangesVersionType determines the preferred version type to use in changes feed entries.
+// If the requested version type is not available, the feed will attempt to fall back to ChangesVersionTypeRevTreeID which should always be available.
+type ChangesVersionType string
+
+const (
+	ChangesVersionTypeRevTreeID ChangesVersionType = "rev" // Use revtree IDs in changes feed entries
+	ChangesVersionTypeCV        ChangesVersionType = "cv"  // Use current version in changes feed entries
+)
+
+func ParseChangesVersionType(s string) (ChangesVersionType, error) {
+	switch ChangesVersionType(s) {
+	case "", ChangesVersionTypeRevTreeID:
+		return ChangesVersionTypeRevTreeID, nil
+	case ChangesVersionTypeCV:
+		return ChangesVersionTypeCV, nil
+	default:
+		return "", fmt.Errorf("unknown changes version type: %q", s)
+	}
+}
+
+// ChangeVersionString attempts to return the version string for the preferred ChangesVersionType, but will fall back to rev if cv is not available when requested.
+func (ce *ChangeEntry) ChangeVersionString(versionType ChangesVersionType) string {
+	if s, ok := ce.Changes[0][versionType]; ok {
+		return s
+	}
+	// requested version type not found, return `rev` as a fallback.
+	return ce.Changes[0][ChangesVersionTypeRevTreeID]
 }
 
 // A changes entry; Database.GetChanges returns an array of these.
 // Marshals into the standard CouchDB _changes format.
 type ChangeEntry struct {
-	Seq            SequenceID      `json:"seq"`
-	ID             string          `json:"id"`
-	Deleted        bool            `json:"deleted,omitempty"`
-	Removed        base.Set        `json:"removed,omitempty"`
-	Doc            json.RawMessage `json:"doc,omitempty"`
-	Changes        []ChangeRev     `json:"changes"`
-	Err            error           `json:"err,omitempty"` // Used to notify feed consumer of errors
-	allRemoved     bool            // Flag to track whether an entry is a removal in all channels visible to the user.
+	Seq            SequenceID            `json:"seq"`
+	ID             string                `json:"id"`
+	Deleted        bool                  `json:"deleted,omitempty"`
+	Removed        base.Set              `json:"removed,omitempty"`
+	Doc            json.RawMessage       `json:"doc,omitempty"`
+	Changes        []ChangeByVersionType `json:"changes"`
```

Comment:
nit:

Not a strong preference here but is `ChangeVersion` a bad name here? to match `DocVersion` etc.

Reply:
I think this is more accurately described as `ChangesByChangesVersionType` but that feels very awkward. Any suggestions?

Reply:
Keeping as `ChangeByVersionType`, since I think it makes it clearer the map keys inside the slice are denoting what version type is being used for the change.

`ChangeByVersion` makes me think the versions are the map keys.